### PR TITLE
Extensions for Akka typed

### PR DESCRIPTION
**Pull request by johanandren**
Refs #16738

Pretty much a straight port from untyped with some minor changes.

### CODE REVIEW

**Comment by johanandren** [comment]
Missing now is test coverage from Java to see that the API works there as well (can't see why it shouldn't)

File: akka-typed/src/main/scala/akka/typed/internal/ExtensionsImpl.scala [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+import java.util.concurrent.{ ConcurrentHashMap, CountDownLatch }
+
+import akka.typed.{ ActorSystem, Extension, ExtensionId, Extensions }
+
+import scala.annotation.tailrec
+import scala.util.{ Failure, Success }
+import scala.collection.JavaConverters._
+
+trait ExtensionsImpl extends Extensions { self: ActorSystem[_] ⇒
```
Comment (johanandren) [inline akka-typed/src/main/scala/akka/typed/internal/ExtensionsImpl.scala]: should be marked internal

File: akka-typed/src/main/scala/akka/typed/Extensions.scala [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+ * }}}
+ *
+ * @tparam T The concrete extension type
+ */
+abstract class ExtensionId[T <: Extension] {
+
+  def createExtension(system: ActorSystem[_]): T
+
+  final def apply(system: ActorSystem[_]): T = system.registerExtension(this)
+  final def create(system: ActorSystem[_]): T = apply(system)
```
Comment (patriknw) [inline akka-typed/src/main/scala/akka/typed/Extensions.scala]: I think `get` is a better name, even though it's lazy get or create.

There is probably some scaladoc you can steal also

File: akka-typed/src/main/scala/akka/typed/internal/ExtensionsImpl.scala [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        }
+      }
+    }
+
+    // eager initialization of CoordinatedShutdown
+    // TODO coordinated shutdown for akka typed
+    // CoordinatedShutdown(self)
+
+    // TODO not sure if we want to keep this one?
+    // loadExtensions("akka.typed.library-extensions", throwOnLoadFail = true)
```
Comment (patriknw) [inline akka-typed/src/main/scala/akka/typed/internal/ExtensionsImpl.scala]: the primary use case was the automatically loaded diagnostics recorder

**Review by patriknw — COMMENTED** [review]
looking good

File: akka-typed/src/main/scala/akka/typed/internal/ExtensionsImpl.scala (hunk 3 above)
Comment (johanandren) [inline akka-typed/src/main/scala/akka/typed/internal/ExtensionsImpl.scala]: Dang, I wrote it and I forgot about it :D Will re-add.

File: akka-typed/src/main/scala/akka/typed/Extensions.scala (hunk 2 above)
Comment (johanandren) [inline akka-typed/src/main/scala/akka/typed/Extensions.scala]: Yep, I realised that when I was working on it now, renamed back to get.

File: akka-typed-tests/src/test/java/akka/typed/ExtensionsTest.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+  public static class MyExtension extends ExtensionId<MyExtImpl> {
+
+    private final static MyExtension instance = new MyExtension();
+
+    private MyExtension() {
+    }
+
+    public static MyExtension getInstance() {
+      return instance;
+    }
```
Comment (johanandren) [inline akka-typed-tests/src/test/java/akka/typed/ExtensionsTest.java]: Skipping the `ExtensionIdProvider` boilerplate and just say that in Scala - use `object` for extensions that can be loaded through config, and for Java - provide a singleton accessor which we call using reflection (and we are anyways doing reflections trixery in the loading of extensions from config).

File: akka-typed-tests/src/test/java/akka/typed/ExtensionsTest.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        Optional.of(ConfigFactory.parseString("akka.typed.extensions += \"akka.typed.ExtensionsTest$MyExtension\"").resolve()),
+        Optional.empty(),
+        Optional.empty()
+    );
+
+    // note that this is not the intended end user way to access it
+    assertTrue(system.hasExtension(MyExtension.getInstance()));
+
+    MyExtImpl instance1 = MyExtension.getInstance().get(system);
+    MyExtImpl instance2 = MyExtension.getInstance().get(system);
```
Comment (johanandren) [inline akka-typed-tests/src/test/java/akka/typed/ExtensionsTest.java]: Ofc a static get is nicer, but would have to be handwritten for Java so I skipped that here.

**Comment by johanandren** [comment]
Weird, jenkins says 5 test failures, but I cannot find any in the output.

**Comment by johanandren** [comment]
PLS BUILD

**Comment by johanandren** [comment]
"did you try turning it off and on again"

File: akka-typed/src/main/scala/akka/typed/Extensions.scala [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+ * public class MyExt extends AbstractExtensionId<MyExtImpl> {
+ *   // single instance of the identifier
+ *   private final static MyExt instance = new MyExt();
+ *
+ *   // protect against other instances than the singleton
+ *   private MyExt() {}
+ *
+ *   // This static method singleton accessor is needed to be able to enable the extension through config when
+ *   // implementing extensions in Java.
+ *   public static MyExt getInstance() {
```
Comment (patriknw) [inline akka-typed/src/main/scala/akka/typed/Extensions.scala]: ah, I see in [docs](http://doc.akka.io/docs/akka/2.5/java/extending-akka.html) that it was not very intuitive from java. I think this convention is good. Document is even more clear that the method must `static` and named `getInstance`. Write that also outside of this code example.

File: akka-typed/src/main/scala/akka/typed/internal/ExtensionsImpl.scala [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+import java.util.concurrent.{ ConcurrentHashMap, CountDownLatch }
+
+import akka.typed.{ ActorSystem, Extension, ExtensionId, Extensions }
+
+import scala.annotation.tailrec
+import scala.util.{ Failure, Success, Try }
+import scala.collection.JavaConverters._
+
+trait ExtensionsImpl extends Extensions { self: ActorSystem[_] ⇒
```
Comment (patriknw) [inline akka-typed/src/main/scala/akka/typed/internal/ExtensionsImpl.scala]: internal api markers

**Review by patriknw — APPROVED** [review]
LGTM after minor adjustments

**Pull request merged**